src/include/duckdb/main/profiling_node.hpp

```diff
@@ -0,0 +1,107 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/main/profiling_node.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/common.hpp"
+#include "duckdb/common/deque.hpp"
+#include "duckdb/common/enums/profiler_format.hpp"
+#include "duckdb/common/pair.hpp"
+#include "duckdb/common/profiler.hpp"
+#include "duckdb/common/reference_map.hpp"
+#include "duckdb/common/string_util.hpp"
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/winapi.hpp"
+#include "duckdb/execution/expression_executor_state.hpp"
+#include "duckdb/execution/physical_operator.hpp"
+#include "duckdb/main/profiling_info.hpp"
+
+namespace duckdb {
+
+enum class ProfilingNodeType : uint8_t { QUERY_ROOT, OPERATOR };
+
+class QueryProfilingNode;
+
+// Recursive tree that mirrors the operator tree
+class ProfilingNode {
+public:
+	virtual ~ProfilingNode() {};
+
+private:
+	ProfilingInfo profiling_info;
+	vector<unique_ptr<ProfilingNode>> children;
+
+	friend class TreeChildrenIterator;
+
+public:
+	idx_t depth = 0;
+	ProfilingNodeType node_type = ProfilingNodeType::OPERATOR;
+
+public:
+	idx_t GetChildCount() {
+		return children.size();
+	}
+
+	ProfilingInfo &GetProfilingInfo() {
+		return profiling_info;
+	}
+
+	const ProfilingInfo &GetProfilingInfo() const {
+		return profiling_info;
+	}
+
+	ProfilingNode *GetChild(idx_t idx) {
+		return children[idx].get();
+	}
+
+	ProfilingNode *AddChild(unique_ptr<ProfilingNode> child) {
+		children.push_back(std::move(child));
+		return children.back().get();
+	}
```

Comment:
Can we return a reference here? `ProfilingNode &AddChild/GetChild`. If not, we should return `optional_ptr` to enable our memory safety checks.